### ISSUE TITLE
Restore VM

### ISSUE THREAD

**Issue by delorean-dev** [EideardVMR/unraid-easybackup#8]
Do you have any date when we can restore a VM's?

**Comment by EideardVMR** [comment]
Hi, no, no date is known yet. Do you need a restore now? I will be happy to help you.
You just need to tell me whether you have made any changes to the configuration of the VM between the backup you want to restore and now and whether there is a snapshot in the backup version or the current one.

**Comment by delorean-dev** [comment]
Hi. No i dont need to restore at moment . Just to know if you have some date to do that :)

**Comment by EideardVMR** [comment]
OK, good.
So the first thing planned is a new GUI. And the following functions are currently on my list: 
- Individual settings for each VM (compression, snapshot backup/stop->start backup, etc.)
- Individual settings for each container (compression)
- The native zip compression should become standard. 
- Backup of the flash should get a tab.
- Copy the last backup of a VM, a container or the flash to an external data carrier at a specific time.
- Copying and compressing shares for a backup on an external data carrier.
- Restoring
- Backup of VMs whose data carriers are located on different paths. 

Because the lack of restoring is an inconvenience, but a restore without the GUI is also feasible for a beginner, I have pushed this to the back. 
For me personally, it is important to be able to create a backup on an external data carrier so that I can also store the most important data outside the home. (e.g. in the event of a fire)

However, I'm leaving the ticket open until I've implemented this function. And if you need help with anything, just get in touch with me. I'm sure we'll find a solution.

**Comment by delorean-dev** [comment]
I'm honestly pleased with your work and your plan.
In fact, what I think is most important is a simple restore of a VM.
I find it unbelievable that NOBODY in the unraid team doesn't have this function in house.
I currently use a script I found on youtube to restore a vm:

https://www.youtube.com/watch?v=OIQKQPqLUfw&t=1s

But it's not practical, and what I'd like is to have a backup of a vm, and if I change disks or buy another machine I can restore it easily and quickly.

I'm open to making you a donation to speed up this process and put it first on your to do list.

**Comment by EideardVMR** [comment]
I agree with you! I also think restoring a backup is important, so this is also part of the plugin and part of the development plan. 

However, the first priority is to have a secure and functioning backup, even if restoring is somewhat complex, the data is there and can be restored. I am therefore happy to offer individual help as long as there is no restore script. 
The data is stored raw in a zip file and not in a proprietary file system. This makes it easy to unpack (even under Windows) and you can then replace the old/damaged files. The only problem with VMs is the necessary configuration which is in the zip file, a small console command is required here. As you can see, in the worst case it is quite easy to restore. I admit, however, that I should publish instructions on this in the wiki or the readme. I think I will do that soon. 

But there is a lot to consider when restoring. Especially the scenario you mentioned (moving to another machine) is an exception. 

A donation is nice and shows me that people appreciate the work, but I do this in my spare time alongside a full-time job and my self-employment. Both take priority. Unless you're Elon Musk and the donation can feed me for the next few years, I'm afraid I'll have to turn it down. It comes as it comes, but I will try to incorporate it into the current rebuild and constantly adapt my priorities to the necessities. 

But I can promise you that I'll keep working on it and you won't have to wait forever.

**Comment by txwireless** [comment]
> OK, good. So the first thing planned is a new GUI. And the following functions are currently on my list:
> 
> * Individual settings for each VM (compression, snapshot backup/stop->start backup, etc.)
> * Individual settings for each container (compression)
> * The native zip compression should become standard.
> * Backup of the flash should get a tab.
> * Copy the last backup of a VM, a container or the flash to an external data carrier at a specific time.
> * Copying and compressing shares for a backup on an external data carrier.
> * Restoring
> * Backup of VMs whose data carriers are located on different paths.
> 
> Because the lack of restoring is an inconvenience, but a restore without the GUI is also feasible for a beginner, I have pushed this to the back. For me personally, it is important to be able to create a backup on an external data carrier so that I can also store the most important data outside the home. (e.g. in the event of a fire)
> 
> However, I'm leaving the ticket open until I've implemented this function. And if you need help with anything, just get in touch with me. I'm sure we'll find a solution.

I want to echo @delorean-dev comment about how much I appreciate your hard work.  I completely agree with both of you that this really should be a core Unraid function.  I also appreciate the challenge of balancing a full-time job, self-employment and supporting a project on the side.

I read this thread just before I was about to open a new issue re: VM snapshot/backup across different paths.  My specific issue is a vm where there is a core img file which is essentially a static boot image and the data and config lives on two SSDs that I passthrough using the /dev/disk-id/ method.  I'm looking at the code to see if I can't quickly hack together a way to exclude the /dev/disk-id/ items and backup the core img file.  In this case the SSDs have their content backed up separately so I don't need them as part of a VM backup.

I've also had the reverse issue in the past where I had a static core img file and then a data img file where the data/config lives and I only want to backup the secondary img on a regular basis.

In your redesign, if you would consider how you have implemented the docker containers where you can select/unselect various bind points and consider in the VM section where you could do a similar select/unselect of disks for the backup.

I previously used Appdata Backup (https://github.com/Commifreak/unraid-appdata.backup) and VM Backup (https://github.com/jtok/unraid.vmbackup) but I like the one-stop shop approach of your plugin.  Please keep up the good work.

**Comment by EideardVMR** [comment]
Hi, thanks for your input.

I'm not 100% happy with the VM backup yet either!
Backups can only be created if all disks are on the same share. The reason is that it is difficult to take into account the individual configurations of each user and at the same time to create relative paths that allow restoring later. 

I considered it unnecessary to exclude individual vDisks as with the dockers (this function was initially planned), as I thought that if I back up a VM, then I always want to back it up completely, otherwise I back up within the VM.

However, I will include this in the planning. But it's not easy, so it will take some time until I have implemented this.

**Comment by michaellunzer** [comment]
<img width="1285" alt="Screenshot 2024-03-08 at 4 36 36 PM" src="https://github.com/EideardVMR/unraid-easybackup/assets/10967175/0020380c-6798-46fc-90a2-02380a0dca4b">

Hi @EideardVMR I'm new to unRaid and backed up a VM I created using your plugin. I should've read the github page saying that the restore button isn't working right now. 

I updated a mac VM and want to roll back the update. Is there a way I can restore the VM from a snapshot on the commandline or another method? Thanks!

**Comment by EideardVMR** [comment]
Hi Michael,
yes, the restore function is not yet active and will not be available in the version. I am currently working on a new version with a completely new GUI in which it should then be available. But that won't help you now.

I see that your VM does not have a guest agent, so your backup will probably not contain a snapshot, which makes restoring very easy.

Since you are brand new, I have made you a GIF of how this works.
But here are the instructions again in writing:
- shut down the VM. IMPORTANT!
- activate a share for domains so that you can access them via Explorer. That makes it easier for you.
- go to your backup share, for me this is backup_internals/vms
- unpack the ZIP. You can do this directly in the share, or copy the zip to your PC first. It doesn't matter. Easybackup could incorrectly recognize a folder with the same name as a backup, so I recommend copying it.
- The ZIP contains a vdisk1.img or another large file, which is your virtual hard disk.
- Now go to the share in which your current VM exists, there is a large file with the same name.
- Rename this file to *.old or something else (just to be on the safe side)
- Copy the vdisk1.img from the zip to the place where the current disk was located.
- Now you can start the VM again and the backup will be loaded.
- You can delete the *.old if you are sure that everything has worked. You do not have to shut down the VM to do this.

**Error handling:**
It may happen that you cannot rename the current vdisk.img because you lack permissions. This is not due to the share but to the folder authorization. Then you have to adjust the chmod.

In my case like this: chmod -R 777 /mnt/user/domains/FreeRadius/

the -R is for recrusive. the 777 stands for the authorization (in this case, everyone can access).

![Animation](https://github.com/EideardVMR/unraid-easybackup/assets/143615696/056a174e-ba3e-43e5-841a-948f3436acf8)

**Comment by michaellunzer** [comment]
Thank you so much for the detailed instructions! I was able to locate the zip containing the image of the hard drive and replace it in the VM directory

Have a great weekend!

**Comment by gacpac** [comment]
Soo, i did a test where i took my cache drive to upgrade, and i created a backup. now i have a new cache drive with nothing on it. i assume i have to copy back the VM to the right place and somehow be able to pull it back :S

Also found the backup didn't backup an additional disk part of the VM thankgod i didn't dispose of the other drive

**Comment by EideardVMR** [comment]
If your drives are mounted on a share, it doesn't matter where they are located. It is only important that the path to the image within the VM points to the restored image. You can also adjust these paths in the VM itself at a later date. 

What do you mean that no additional hard disk parts were backed up? 

Normally, all hard disks (images) that are mounted should also be backed up. Regardless of whether you have mounted 1,2,3,10 images. I have several systems with several images and they are all backed up. I would be interested to know what you are missing.

**Comment by gacpac** [comment]
yeah so, i have a vm for gns3 ran a backup and it only backed up 1 disk. After i restored everything i tried again to backup and again it only backed up 1 disk. Hae no idea :s 

My bit of a problem restoring was i had the default location backup_internal, all the vms are there. I had the feeling that the plugin was going to find the VMs and from there i could unzip,restore, etc. 

Since the plugin was not finding VMs i had to unzip using krusader (to be faster). then add the VMS. Maybe is the way GNS3 vm is configured that it doesn't like backing up the 2nd disk?

**Comment by EideardVMR** [comment]
OK, so as far as finding the backups is concerned, it really is the case that an automatic restore process does NOT exist. Currently at least!

But only a restore is planned for existing VMs. So if there is an error in the VM. Your point is good, it should also be possible to import existing backups again. I'll add that to the agenda! Thank you!

Once again about the images. Are the hard disks recognized correctly under "Snapshot"? This will not be due to the VM itself! At most a missing guest agent. But then it would not make a backup at all. 

Example: 
![image](https://github.com/EideardVMR/unraid-easybackup/assets/143615696/2ce53568-4e1a-44f1-ab57-49cf1c96905a)

**Comment by gacpac** [comment]
Well, after restored the VMs manually unzipping everything came back up
detected in the plugin.

**Comment by EideardVMR** [comment]
And if you make a backup now, will all images be backed up or just one?
